### PR TITLE
Add manifest file caching support for Iceberg Hive Metastore catalog

### PR DESCRIPTION
**Pull request by hackeryang**
## Description
Iceberg library has a manifest caching feature contributed by Cloudera:  
https://blog.cloudera.com/12-times-faster-query-planning-with-iceberg-manifest-caching-in-impala/  

PrestoDB also has similar configurations to turn on this feature: https://github.com/prestodb/presto/pull/21399  

Iceberg manifest caching can accelerate query planning.

## Additional context and related issues
https://prestodb.io/docs/current/connector/iceberg.html#manifest-file-caching  

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Add manifest file caching support for Iceberg Hive Metastore catalog. ({issue}`22376`)
```

### CODE REVIEW

File: docs/src/main/sphinx/connector/iceberg.md [hunk 1]
```diff
@@ -1687,3 +1687,36 @@ The connector supports redirection from Iceberg tables to Hive tables with the
...
+* - `iceberg.io.manifest.cache-enabled`
+  - Enable or disable the manifest caching feature.
+  - `false`
+* - `iceberg.io.manifest.cache.max-total-size`
+  - Maximum size of cache size.
+  - `100MB`
+* - `iceberg.io.manifest.cache.expiration-interval-duration`
+  - Maximum time duration for which an entry stays in the manifest cache.
+  - `60s`
+* - `iceberg.io.manifest.cache.max-content-length`
```
Comment (findepi) [inline docs/src/main/sphinx/connector/iceberg.md:1718]: I am not convinced we need "io" in the config property names.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/snowflake/SnowflakeIcebergTableOperationsProvider.java [hunk 2]
```diff
@@ -13,32 +13,39 @@
...
 
+import static io.trino.plugin.iceberg.IcebergUtil.loadManifestCachingProperties;
 import static java.util.Objects.requireNonNull;
 
 public class SnowflakeIcebergTableOperationsProvider
         implements IcebergTableOperationsProvider
 {
     private final String snowflakeDatabase;
     private final TrinoFileSystemFactory fileSystemFactory;
+    private final IcebergConfig icebergConfig;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/snowflake/SnowflakeIcebergTableOperationsProvider.java:38]: IcebergConfig is a mutable. we avoid shared mutable state. please do not keep the config on the field, just extract what you need from it in the constructor.

File: docs/src/main/sphinx/connector/iceberg.md [hunk 3]
```diff
@@ -1687,3 +1687,36 @@ The connector supports redirection from Iceberg tables to Hive tables with the
...
 caching](/object-storage/file-system-cache).
+
+### Manifest File Caching
+
+As of Iceberg version 1.1.0, Apache Iceberg provides a mechanism to cache the
+contents of Iceberg manifest files in memory. This feature helps to reduce
+repeated reads of small Iceberg manifest files from remote storage.
+
+Note that currently, manifest file caching is supported for Hive Metastore, Glue,
+JDBC, Nessie, Rest and Snowflake catalogs in the Iceberg connector.
```
Comment (findepi) [inline docs/src/main/sphinx/connector/iceberg.md:1698]: If the feature requires a catalog specific code, it should not be configured in IcebergConfig, but rather a separate (new) config class that catalog modules will bind or not, depending whether they support given feature.

this prevents a situation where config is bound & available, but not used, doesn't mean anything.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 4]
```diff
@@ -954,4 +958,13 @@ public static long getModificationTime(String path, TrinoFileSystem fileSystem)
             throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed to get file modification time: " + path, e);
         }
     }
+
+    public static Map<String, String> loadManifestCachingProperties(Map<String, String> properties, IcebergConfig icebergConfig)
+    {
+        properties.put(IO_MANIFEST_CACHE_ENABLED, "true");
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:964]: this is configuring `org.apache.iceberg.ManifestFiles#CONTENT_CACHES`, right?
the cache is keyed by `FileIO` and our `ForwardingFileIo` is typically short-lived. how effective is this caching?

cc @alexjo2144 @electrum

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java [hunk 5]
```diff
@@ -430,6 +439,58 @@ public IcebergConfig setSplitManagerThreads(int splitManagerThreads)
...
+        return this;
+    }
+
+    public Duration getManifestCacheExpireDuration()
+    {
+        return manifestCacheExpireDuration;
+    }
+
+    @Config("iceberg.io.manifest.cache.expiration-interval-duration")
+    @ConfigDescription("Maximum duration for which an entry stays in the manifest cache")
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:474]: Are manifest files mutable or this is mostly for releasing memory?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 6]
```diff
@@ -954,4 +958,13 @@ public static long getModificationTime(String path, TrinoFileSystem fileSystem)
...
+        properties.put(IO_MANIFEST_CACHE_MAX_TOTAL_BYTES, String.valueOf(icebergConfig.getMaxManifestCacheSize().toBytes()));
```
Comment (sopel39) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:965]: How big is the default?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java (hunk 5 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:474]: immutable

File: docs/src/main/sphinx/connector/iceberg.md (hunk 3 above)
Comment (hackeryang) [reply docs/src/main/sphinx/connector/iceberg.md:1698]: The feature is integrated within the Iceberg library, we just need to pass the configurations to the Iceberg api, so you are right, it isn't a catalog specific code, i will remove this description~

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 4 above)
Comment (hackeryang) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:964]: Yes, in one of our production environments, a big sql which was planned for 12s has been reduced to 8s~

In our internal implementation, we also used a Caffeine cache with the key `session user` and the value `FileIo` to make a `ForwardingFileIo` live longer, i can also add relevant codes to our PR

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 6 above)
Comment (hackeryang) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:965]: By default it is 100MB, integrated in the Iceberg library:  
image

File: docs/src/main/sphinx/connector/iceberg.md (hunk 3 above)
Comment (findepi) [reply docs/src/main/sphinx/connector/iceberg.md:1698]: the plumbing is catalog-specific, afaict from reading the code

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:964]: the 12s -> 8s improvement -- was it with or without fileio caching?

what would the numbers be if we don't do this PR and instead enable filesystem caching that we already have in trino?

File: docs/src/main/sphinx/connector/iceberg.md (hunk 3 above)
Comment (hackeryang) [reply docs/src/main/sphinx/connector/iceberg.md:1698]: > the plumbing is catalog-specific, afaict from reading the code

You are right, after reading the iceberg codes: https://github.com/apache/iceberg/pull/4518/files  
I found that it only support `HadoopFileIO`

**Comment by findinpath** [comment]
@hackeryang  pls try adding a test into `io.trino.plugin.iceberg.TestIcebergFileOperations`

I had taken a stint on integrating this functionality into Iceberg as well some time ago in https://github.com/trinodb/trino/pull/20862 and the limitation on which I've stumbled was 

> At the time of this writing, the main limitation in Trino to use this feature is that each time a storage operation is done, a new `ForwardingFileIo` instance is being created, while the `CONTENT_CACHES` static field from Iceberg awaits the same `FileIO` instance to retrieve a certain content cache.

I'm hopeful though that I missed something in my original PR and that your feature is effective for making use of the manifest cache.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 4 above)
Comment (hackeryang) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:975]: > the 12s -> 8s improvement -- was it with or without fileio caching?
> 
> what would the numbers be if we don't do this PR and instead enable filesystem caching that we already have in trino?

@findepi the 12s -> 8s improvement was with fileio/manifest caching.  

Sorry our newest production environment is 423, so i couldn't test the performance about filesystem caching, but i saw that `PrestoDB` have `manifest caching` and `filesystem caching` both, so i thought that they may not conflict with each other~ The metadata caching in memory is also faster than caching in disks in most cases.  

Thank you for your advice, I have modified some codes accordingly, please review again when you have time~

**Comment by hackeryang** [comment]
> @hackeryang pls try adding a test into `io.trino.plugin.iceberg.TestIcebergFileOperations`
> 
> I had taken a stint on integrating this functionality into Iceberg as well some time ago in #20862 and the limitation on which I've stumbled was
> 
> > At the time of this writing, the main limitation in Trino to use this feature is that each time a storage operation is done, a new `ForwardingFileIo` instance is being created, while the `CONTENT_CACHES` static field from Iceberg awaits the same `FileIO` instance to retrieve a certain content cache.
> 
> I'm hopeful though that I missed something in my original PR and that your feature is effective for making use of the manifest cache.

We used a Caffeine cache in `HiveMetastoreTableOperationsProvider` with the key `CatalogType` and the value `FileIO` to reduce the frequency of temporary creations of fileio~ 

Sorry i haven't figure out good logic and APIs to write the ut yet, i will try to add one if i have a better idea

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:975]: > @findepi the 12s -> 8s improvement was with fileio/manifest caching.

thanks!

before merging this PR, we should understand what's the level of improvement that it brings (in isolation).

cc @sopel39 @raunaqmorarka

**Comment by mosabua** [comment]
fyi @cwsteinbach .. maybe you can help here.

**Comment by raunaqmorarka** [comment]
I have another PR adding iceberg metadata caching through a different approach https://github.com/trinodb/trino/pull/22739
That PR avoids depending on iceberg's manifest file caching and just implements the caching within Trino. I had attempted using iceberg manifest file caching as well, it forced changes around lifecycle of FileIO object because of the caching being tied to FileIO instance in iceberg. It's also harder to re-use across queries as we'd have to retain FileIO objects across queries to make that work and that may not be appropriate each of those objects may have different `ConnectorIdentity` per query. Implementing it in Trino also makes it possible to potentially reuse in delta connector.

**Comment by mosabua** [comment]
That sounds great @raunaqmorarka - can we conclude that we should therefore close this PR and help with reviews and input on yours?

**Comment by raunaqmorarka** [comment]
> That sounds great @raunaqmorarka - can we conclude that we should therefore close this PR and help with reviews and input on yours?

yup, i think so.
@hackeryang please try out https://github.com/trinodb/trino/pull/22739 and let me know if you run into any problems with that.

**Comment by mosabua** [comment]
Closing as discussed in the previous comments since the approach in this PR is inferior to the linked PR from @raunaqmorarka  - we will continue to collaborate their and would love to get review and testing feedback.

**Pull request closed without merging**